### PR TITLE
415 documentation for winkler interval score

### LINKED ISSUE #415: Documentation for Winkler Interval Score
**Is your documentation request related to a problem? Please describe.**
There is no documentation for the Winkler Interval Score from this issue: #407.

**Describe the solution you'd like**
Add information regarding this score, the formulas used to calculate it and how to interpret it.

### PR DESCRIPTION
**Pull request by LacombeLouis**
# Description
Add description for all the metrics.

Fixes #415

Please remove options that are irrelevant.
- This change requires a documentation update

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [x] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [x] Documentation builds successfully : `make doc`

### CODE REVIEW

File: doc/theoretical_description_metrics.rst [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+
+Regression Mean Width Score
+---------------------------
+
+The **Regression Mean Width Score** assesses the average width of the prediction intervals provided by the model.
+
+.. math::
+
+   \text{Mean Width} = \frac{1}{n} \sum_{i=1}^{n} (y_{\text{pred, up}}^{(i)} - y_{\text{pred, low}}^{(i)})
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:38]:
```suggestion
   \text{Mean Width} = \frac{1}{n} \sum_{i=1}^{n} (\hat y^{\text{up}}_{i} - \hat y^{\text{low}}_{i})
```

File: doc/theoretical_description_metrics.rst [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+.. math::
+
+   C = \frac{1}{n} \sum_{i=1}^{n} \mathbf{1}(y_{\text{pred, low}}^{(i)} \leq y_{\text{true}}^{(i)} \leq y_{\text{pred, up}}^{(i)})
+
+where:
+
+- :math:`n` is the number of samples,
+- :math:`y_{\text{true}}^{(i)}` is the true value for the :math:`i`-th sample,
+- :math:`y_{\text{pred, low}}^{(i)}` and :math:`y_{\text{pred, up}}^{(i)}` are the lower and upper bounds of the prediction intervals, respectively.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:28]:
```suggestion
- :math:`n` is the number of samples,
- :math:`y_{i}` is the true value for the :math:`i`-th sample,
- :math:`\hat y^{\text{low}}_{i}` and :math:`\hat y^{\text{up}}_{i}` are the lower and upper bounds of the prediction intervals, respectively.
```

File: doc/theoretical_description_metrics.rst [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+Classification Mean Width Score
+-------------------------------
+
+For classification tasks, the **Classification Mean Width Score** calculates the average size of the prediction sets across all samples.
+
+.. math::
+
+   \text{Mean Width} = \frac{1}{n} \sum_{i=1}^{n} |\text{Set}_{\text{pred}}^{(i)}|
+
+where :math:`|\text{Set}_{\text{pred}}^{(i)}|` denotes the number of classes included in the prediction set for sample :math:`i`.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:62]:
```suggestion
   \text{Mean Width} = \frac{1}{n} \sum_{i=1}^{n} |\hat C_{x_i}|

where :math:`|\hat C_{x_i}|` denotes the number of classes included in the prediction set for sample :math:`i`.
```

File: doc/theoretical_description_metrics.rst [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+Size-Stratified Coverage (SSC)
+-------------------------------
+
+**Size-Stratified Coverage (SSC)** evaluates how the size of prediction sets or intervals affects their ability to cover the true outcomes [1]. It's calculated separately for classification and regression:
+
+**Regression:**
+
+.. math::
+
+   \text{SSC}_{\text{regression}} = \sum_{k=1}^{K} \left( \frac{1}{|I_k|} \sum_{i \in I_k} \mathbf{1}(y_{\text{pred, low}}^{(i)} \leq y_{\text{true}}^{(i)} \leq y_{\text{pred, up}}^{(i)}) \right)
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:74]:
```suggestion
   \text{SSC}_{\text{regression}} = \sum_{k=1}^{K} \left( \frac{1}{|I_k|} \sum_{i \in I_k} \mathbf{1}(\hat y^{\text{low}}_{i} \leq y_{i} \leq \hat y^{\text{up}}_{i}) \right)
```

File: doc/theoretical_description_metrics.rst [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+**Classification:**
+
+.. math::
+
+   \text{SSC}_{\text{classification}} = \sum_{k=1}^{K} \left( \frac{1}{|S_k|} \sum_{i \in S_k} \mathbf{1}(y_{\text{true}}^{(i)} \in \text{Set}_{\text{pred}}^{(i)}) \right)
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:80]:
```suggestion
   \text{SSC}_{\text{classification}} = \sum_{k=1}^{K} \left( \frac{1}{|S_k|} \sum_{i \in S_k} \mathbf{1}(y_{i} \in \hat C(x_i)) \right)
```

File: doc/theoretical_description_metrics.rst [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+The **Coverage Width-Based Criterion (CWC)** evaluates prediction intervals by balancing their empirical coverage and width. It is designed to both reward narrow intervals and penalize those that do not achieve a specified coverage probability [6].
+
+.. math::
+
+   \text{CWC} = (1 - \text{Mean Width Score}) \times \exp\left(-\eta \times (\text{Coverage Score} - (1-\alpha))^2\right)
+
+
+
+Regression MWI Score
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:116]:
```suggestion
Mean Winkler Interval Score
```

File: doc/theoretical_description_metrics.rst [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+--------------------
+
+The **Regression MWI (Mean Winkler Interval) Score** evaluates prediction intervals by combining their width with a penalty for intervals that do not contain the observation [8, 10].
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:119]:
```suggestion
The **MWI (Mean Winkler Interval) Score** evaluates prediction intervals by combining their width with a penalty for intervals that do not contain the observation [8, 10].
```

File: doc/theoretical_description_metrics.rst [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+.. math::
+
+   \text{MWI Score} = \frac{1}{n} \sum_{i=1}^{n} (y_{\text{pred, up}}^{(i)} - y_{\text{pred, low}}^{(i)}) + \frac{2}{\alpha} \sum_{i=1}^{n} \max(0, |y_{\text{true}}^{(i)} - y_{\text{pred, boundary}}^{(i)}|)
+
+where :math:`y_{\text{pred, boundary}}^{(i)}` is the nearest interval boundary not containing :math:`y_{\text{true}}^{(i)}`, and :math:`\alpha` is the significance level.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:125]:
```suggestion
   \text{MWI Score} = \frac{1}{n} \sum_{i=1}^{n} (\hat y^{\text{up}}_{i} - \hat y^{\text{low}}_{i}) + \frac{2}{\alpha} \sum_{i=1}^{n} \max(0, |y_{i} - \hat y^{\text{boundary}}_{i}|)

where :math:`\hat y^{\text{boundary}}_{i}` is the nearest interval boundary not containing :math:`y_{i}`, and :math:`\alpha` is the significance level.
```

File: doc/theoretical_description_metrics.rst [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+
+Cumulative Differences
+----------------------
+
+**Cumulative Differences** calculates the cumulative differences between sorted true values and prediction scores, helping to understand how well the prediction scores correspond to the actual outcomes when both are ordered by the score [2].
+
+.. math::
+
+   \text{Cumulative Differences} = \frac{1}{n} \sum_{i=1}^{n} (y_{\text{true,sorted}}^{(i)} - y_{\text{score,sorted}}^{(i)})
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:175]:
```suggestion
   \text{Cumulative Differences} = \frac{1}{n} \sum_{i=1}^{n} (y_{\sigma_1(i)} - \hat y_{\sigma_2(i)})

where:

- :math:`\sigma_1` is the permutation which sorts all the true values.
- :math:`\sigma_2` is the permutation which sorts all the predicted values.
```

File: doc/theoretical_description_metrics.rst [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+
+Spiegelhalter’s Test
+--------------------
+
+**Spiegelhalter’s Test** assesses the calibration of binary predictions based on a transformation of the Brier score [9].
+
+.. math::
+
+   \text{Spiegelhalter's Statistic} = \frac{\sum (y_{\text{true}} - y_{\text{score}})(1 - 2y_{\text{score}})}{\sqrt{\sum (1 - 2y_{\text{score}})^2 y_{\text{score}} (1 - y_{\text{score}})}}
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:207]:
```suggestion
   \text{Spiegelhalter's Statistic} = \frac{\sum_{i=1}^n (y_i - \hat y_i)(1 - 2\hat y_i)}{\sqrt{\sum_{i=1}^n (1 - 2 \hat y_i)^2 \hat y_i (1 - \hat y_i)}}
```

File: doc/theoretical_description_metrics.rst [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+
+
+Top-Label Expected Calibration Error (Top-Label ECE)
+----------------------------------------------------
+
+**Top-Label ECE** focuses on the class predicted with the highest confidence for each sample, assessing whether these top-predicted confidences align well with actual outcomes. It is calculated by dividing the confidence score range into bins and comparing the mean confidence against empirical accuracy within these bins [5].
+
+.. math::
+
+   \text{Top-Label ECE} = \sum_{b=1}^{B} \frac{n_b}{n} \left| \text{acc}_b - \text{conf}_b \right|
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:156]: The difference with the previous metric is not clear (the equations are too similar).

**Review by thibaultcordier — CHANGES_REQUESTED** [review]
Wow, this detailed page explaining MAPIE metrics is awesome. Thank you so much for listing everything here.

I've commented on the style and a question about the equations being too similar.

File: doc/theoretical_description_metrics.rst (hunk 11 above)
Comment (LacombeLouis) [reply doc/theoretical_description_metrics.rst]: Indeed, I think I copy-pasted the same one. Let me update this.

File: doc/theoretical_description_metrics.rst [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+
+This document provides detailed descriptions of various metrics used to evaluate the performance of predictive models, particularly focusing on their ability to estimate uncertainties and calibrate predictions accurately.
+
+1. General Metrics
+==================
+
+Regression Coverage Score
+-------------------------
+
+The **Regression Coverage Score** calculates the fraction of true outcomes that fall within the provided prediction intervals.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:16]:
```suggestion
The **Regression Coverage Score (RCS)** calculates the fraction of true outcomes that fall within the provided prediction intervals.
```

File: doc/theoretical_description_metrics.rst [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+where:
+
+- :math:`n` is the number of samples,
+- :math:`y_{i}` is the true value for the :math:`i`-th sample,
+- :math:`\hat y^{\text{low}}_{i}` and :math:`\hat y^{\text{up}}_{i}` are the lower and upper bounds of the prediction intervals, respectively.
+
+Regression Mean Width Score
+---------------------------
+
+The **Regression Mean Width Score** assesses the average width of the prediction intervals provided by the model.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:31]:
```suggestion
The **Regression Mean Width Score (RMWS)** assesses the average width of the prediction intervals provided by the model.
```

File: doc/theoretical_description_metrics.rst [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+
+.. math::
+
+   \text{Mean Width} = \frac{1}{n} \sum_{i=1}^{n} (\hat y^{\text{up}}_{i} - \hat y^{\text{low}}_{i})
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:35]:
```suggestion
   \text{RMWS} = \frac{1}{n} \sum_{i=1}^{n} (\hat y^{\text{up}}_{i} - \hat y^{\text{low}}_{i})
```

File: doc/theoretical_description_metrics.rst [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+
+Classification Coverage Score
+-----------------------------
+
+The **Classification Coverage Score** measures how often the true class labels fall within the predicted sets.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:40]:
```suggestion
The **Classification Coverage Score (CCS)** measures how often the true class labels fall within the predicted sets.
```

File: doc/theoretical_description_metrics.rst [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+Classification Mean Width Score
+-------------------------------
+
+For classification tasks, the **Classification Mean Width Score** calculates the average size of the prediction sets across all samples.
+
+.. math::
+
+   \text{Mean Width} = \frac{1}{n} \sum_{i=1}^{n} |\hat C_{x_i}|
+
+where :math:`|\hat C_{x_i}|` denotes the number of classes included in the prediction set for sample :math:`i`.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:57]:
```suggestion
where :math:`|\hat C(x_i)|` denotes the number of classes included in the prediction set for sample :math:`i`.
```

File: doc/theoretical_description_metrics.rst [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+Here, :math:`\hat C(x_{i})` represents the set of predicted labels that could possibly contain the true label for the :math:`i`-th observation :math:`x_{i}`.
+
+Classification Mean Width Score
+-------------------------------
+
+For classification tasks, the **Classification Mean Width Score** calculates the average size of the prediction sets across all samples.
+
+.. math::
+
+   \text{Mean Width} = \frac{1}{n} \sum_{i=1}^{n} |\hat C_{x_i}|
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:55]:
```suggestion
   \text{CMWS} = \frac{1}{n} \sum_{i=1}^{n} |\hat C(x_i)|
```

File: doc/theoretical_description_metrics.rst [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+.. math::
+
+   CCS = \frac{1}{n} \sum_{i=1}^{n} \mathbf{1}(y_{i} \in \hat C(x_{i}))
+
+Here, :math:`\hat C(x_{i})` represents the set of predicted labels that could possibly contain the true label for the :math:`i`-th observation :math:`x_{i}`.
+
+Classification Mean Width Score
+-------------------------------
+
+For classification tasks, the **Classification Mean Width Score** calculates the average size of the prediction sets across all samples.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:51]:
```suggestion
For classification tasks, the **Classification Mean Width Score (CMWS)** calculates the average size of the prediction sets across all samples.
```

File: doc/theoretical_description_metrics.rst [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+
+.. math::
+
+   \text{Mean Width} = \frac{1}{n} \sum_{i=1}^{n} |\hat C_{x_i}|
+
+where :math:`|\hat C_{x_i}|` denotes the number of classes included in the prediction set for sample :math:`i`.
+
+Size-Stratified Coverage (SSC)
+-------------------------------
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:60]:
```suggestion
Size-Stratified Coverage
-------------------------
```

File: doc/theoretical_description_metrics.rst [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+
+   \text{SSC}_{\text{classification}} = \sum_{k=1}^{K} \left( \frac{1}{|S_k|} \sum_{i \in S_k} \mathbf{1}(y_{i} \in \hat C(x_i)) \right)
+
+where:
+
+- :math:`K` is the number of distinct size groups,
+- :math:`I_k` and :math:`S_k` are the indices of samples whose prediction intervals or sets belong to the :math:`k`-th size group.
+
+Hilbert-Schmidt Independence Criterion (HSIC)
+----------------------------------------------
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:82]:
```suggestion
Hilbert-Schmidt Independence Criterion
---------------------------------------
```

File: doc/theoretical_description_metrics.rst [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+
+where:
+
+- :math:`\mathbf{K}` and :math:`\mathbf{L}` are the kernel matrices representing the interval sizes and coverage indicators, respectively.
+- :math:`\mathbf{H}` is the centering matrix, :math:`\mathbf{H} = \mathbf{I} - \frac{1}{n} \mathbf{11}^\top`.
+
+This measure is crucial for determining whether certain sizes of prediction intervals are systematically more or less likely to contain the true values, which can highlight biases in interval-based predictions.
+
+Coverage Width-Based Criterion (CWC)
+------------------------------------
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:98]:
```suggestion
Coverage Width-Based Criterion
------------------------------
```

File: doc/theoretical_description_metrics.rst [hunk 22]
```diff
@@ -0,0 +1,255 @@
...
+
+   \text{MWI Score} = \frac{1}{n} \sum_{i=1}^{n} (\hat y^{\text{up}}_{i} - \hat y^{\text{low}}_{i}) + \frac{2}{\alpha} \sum_{i=1}^{n} \max(0, |y_{i} - \hat y^{\text{boundary}}_{i}|)
+
+where :math:`\hat y^{\text{boundary}}_{i}` is the nearest interval boundary not containing :math:`y_{i}`, and :math:`\alpha` is the significance level.
+
+2. Calibration Metrics
+======================
+
+Expected Calibration Error (ECE)
+--------------------------------
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:121]:
```suggestion
Expected Calibration Error
--------------------------
```

**Review by thibaultcordier — CHANGES_REQUESTED** [review]
Thank you for your changes. I just need to add a few harmonised changes. At the same time, I can't see your changes in the documentation (bug because there are a lot of builds / a lot of commits). I suggest you make a single commit for all the suggested changes with the "Add suggestion to batch" option in GitHub.

File: doc/theoretical_description_metrics.rst (hunk 11 above)
Comment (LacombeLouis) [reply doc/theoretical_description_metrics.rst]: done

File: doc/theoretical_description_metrics.rst [hunk 23]
```diff
@@ -0,0 +1,255 @@
...
+Expected Calibration Error
+--------------------------
+
+The **Expected Calibration Error** (ECE) is a metric used to evaluate how well the predicted probabilities of a model align with the actual outcomes. The ECE provides a measure of the difference between predicted confidence levels and actual accuracy. The idea is to divide the predictions into bins based on confidence scores and then compare the accuracy within each bin to the average confidence level of the predictions in that bin.
+
+.. math::
+    \text{ECE} = \sum_{i=1}^B \frac{|B_i|}{n} \left| \text{acc}(B_i) - \text{conf}(B_i) \right|
+
+where:
+- :math:`B_i` is the set of indices of samples that fall into the i-th bin.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:129]:
```suggestion
where:

- :math:`B_i` is the set of indices of samples that fall into the i-th bin.
```

File: doc/theoretical_description_metrics.rst [hunk 24]
```diff
@@ -0,0 +1,255 @@
...
+
+The **Top-Label Expected Calibration Error** (Top-Label ECE) extends the concept of ECE to the multi-class setting. Instead of evaluating calibration over all predicted probabilities, Top-Label ECE focuses on the calibration of the most confident prediction (top-label) for each sample.
+
+The Top-Label ECE is calculated as follows:
+
+.. math::
+    \text{Top-Label ECE} = \frac{1}{L} \sum_{j=1}^L \sum_{i=1}^B \frac{|B_{i,j}|}{n_j} \left| \text{acc}(B_{i,j}) - \text{conf}(B_{i,j}) \right|
+
+where:
+- :math:`L` is the number of unique labels.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:149]:
```suggestion
where:

- :math:`L` is the number of unique labels.
```

**Review by thibaultcordier — CHANGES_REQUESTED** [review]
Add line breaks, and it will be perfect.

File: examples/regression/4-tutorials/plot_main-tutorial-regression.py [hunk 25]
```diff
@@ -1,32 +1,25 @@
-"""
+r"""
 ===============================
 Tutorial for tabular regression
 ===============================
+
+In this tutorial, we compare the prediction intervals estimated by MAPIE on a
+simple, one-dimensional, ground truth function :math:`f(x) = x \times \sin(x)`.
+Throughout this tutorial, we will answer the following questions:
+- How well do the MAPIE strategies capture the aleatoric uncertainty
```
Comment (thibaultcordier) [inline examples/regression/4-tutorials/plot_main-tutorial-regression.py:9]: Add a line break here:
```suggestion
Throughout this tutorial, we will answer the following questions:

- How well do the MAPIE strategies capture the aleatoric uncertainty
```

File: doc/theoretical_description_metrics.rst [hunk 26]
```diff
@@ -0,0 +1,316 @@
...
+    \text{conf}(B_m) = \frac{1}{\left| B_m \right|} \sum_{i \in B_m} \hat{f}(x_i)
+
+The ECE is then calculated using the following formula:
+
+.. math::
+
+    \text{ECE} = \sum_{m=1}^M \frac{\left| B_m \right|}{n} \left| \text{acc}(B_m) - \text{conf}(B_m) \right|
+
+where:
+- :math:`B_m` is the set of indices of samples that fall into the :math:`m`-th bin.
```
Comment (thibaultcordier) [inline doc/theoretical_description_metrics.rst:144]: Add line breack here:
```suggestion
where:

- :math:`B_m` is the set of indices of samples that fall into the :math:`m`-th bin.
```

**Review by thibaultcordier — CHANGES_REQUESTED** [review]
Simply add a line break to the format and you're done.

**Review by thibaultcordier — APPROVED** [review]

**Pull request merged**